jgrapht-core/src/main/java/org/jgrapht/alg/shortestpath/YenKShortestPathsIterator.java

```diff
@@ -0,0 +1,283 @@
+                                                                 // returns
+                                                                 // empty
+                                                                 // path
+                                                                 // instead
+                                                                 // of null
+                // Entire path is made up of the root path and spur path.
+                ArrayList<E> totalPath = new ArrayList<E>();
+                Iterator<V> iter;
+                V first, current;
+                double weight = 0.0;
+                if (!rootPath.isEmpty()) {
+                    iter = rootPath.iterator();
+                    first = iter.next();
+                    current = first;
+                    while (iter.hasNext()) {
+                        V v = iter.next();
+                        E e = graph.getEdge(current, v);
+                        weight += graph.getEdgeWeight(e);
+                        totalPath.add(e);
+                        current = v;
+                    }
+                    iter = spurPath.getVertexList().iterator();
+                } else {
+                    iter = spurPath.getVertexList().iterator();
+                    first = iter.next();
+                    current = first;
+                }
+                while (iter.hasNext()) {
+                    V v = iter.next();
+                    E e = graph.getEdge(current, v);
+                    weight += graph.getEdgeWeight(e);
+                    totalPath.add(e);
+                    current = v;
+                }
+                // Add the potential k-shortest path to the heap.
+                GraphWalk<V, E> gw = new GraphWalk<V, E>(graph, first, current, totalPath, weight);
+                b.add(gw);
+            }
+            // Add back the edges and nodes that were removed from the
+            // graph.
+            // restore edges to Graph;
+            // restore nodes in rootPath to Graph;
+            // not needed, we simply throw away the mask.
+        }
+        if (b.isEmpty()) {
+            // This handles the case of there being no spur paths, or no
+            // spur paths left.
+            // This could happen if the spur paths have already been
+            // exhausted (added to A),
+            // or there are no spur paths at all - such as when both the
+            // source and sink vertices
+            // lie along a "dead end".
+            return;
+        }
+        // Sort the potential k-shortest paths by cost.
+        // This is done already at insertion to the PriorityQueue.
+        // Add the lowest cost path becomes the k-shortest path.
+        GraphPath<V, E> csp;
+        do {
+            csp = b.poll();
+            if (null == csp) {
+                return;
+            }
+        } while (lastShortestPath.getEdgeList().equals(csp.getEdgeList()));

```

Comment:
This confuses me. By definition list b should not contain paths which are identical to lastShortestPath, so this loop seems unnecessary? `a.add(b.poll())` should suffice?

Reply:
the null check was indeed redundant. removed it.
the loop is needed.